feat: add a flask_context_manager.

laipz8200 opened the pull request:
Signed-off-by: -LAN- <laipz8200@outlook.com>

> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 2. Ensure there is an associated issue and you have been assigned to it
> 3. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

part of #20449

Replaces manual context management and user restoration logic with a
unified `flask_context_manager` for better code maintainability and
clarity. Removes redundant import of `has_request_context` from
multiple modules. Updates `.gitignore` to exclude `.roo/` directory,
presumably related to AI Assistant outputs.


## Screenshots

| Before | After |
|--------|-------|
| ...    | ...   |

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

laipz8200 committed:
feat: add a flask_context_manager.

Replaces manual context management and user restoration logic with a
unified `flask_context_manager` for better code maintainability and
clarity. Removes redundant import of `has_request_context` from
multiple modules. Updates `.gitignore` to exclude `.roo/` directory,
presumably related to AI Assistant outputs.

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
chore: add blank line

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
chore: Remove unnecessary request context copying in threads

Eliminates the usage of Flask's `copy_current_request_context` decorator
for thread creation across multiple app generators. The change simplifies
thread initialization by directly passing necessary context and parameters
to the worker function. This refactoring reduces complexity and potential
overhead associated with copying request contexts when not required.

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
refactor: Refines flask context manager interface

Simplifies the `flask_context_manager` function by requiring a
`contextvars.Context` object instead of an optional mapping of
`ContextVar` instances. This change enhances type safety and
streamlines the function's usage, ensuring a more consistent
and clear interface for managing context variables in Flask
applications.

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost commented on api/core/helper/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from core.helper.flask_context import flask_context_manager
```
I think we should avoid creating broad packages named `helper` or `utils`, as they tend to become catch-all directories that grow unwieldy over time.

Instead, I'd recommend placing `flask_context_manager` in a more specific location like `libs/flask_utils.py`. This approach offers several benefits:

1. **Better semantic grouping** - Flask-related utilities are clearly organized together
2. **Cleaner imports** - Direct imports are more explicit and easier to track
3. **Improved maintainability** - Avoids the "junk drawer" problem common with generic helper packages

The import would then be:

```python
from libs.flask_utils import flask_context_manager
```

This follows the principle of organizing code by domain/functionality rather than by generic categories.

laipz8200 replied:
SGTM

laipz8200 committed:
chore: Adds note on non-recommended usage of context manager

Introduces a note in the flask_context_manager function documentation to clarify
that while the manager facilitates using current_user across threads and app
contexts, directly passing the user in parameters is the preferred approach.

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost reviewed (dismissed):
Other changes LGTM.

QuantumGhost commented on api/core/helper/flask_context.py:
```diff
@@ -0,0 +1,65 @@
+    are set within the new context.
+
+    Note:
+        This manager aims to allow use current_user cross thread and app context,
```
👍

laipz8200 committed:
refactor: move flask_context_manager to libs.flask_utils

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
test(flask_utils): Add tests for thread safety of Flask current_user

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
refactor: Renames flask_context_manager to preserve_flask_contexts

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost reviewed (approved):
LGTM
